scanpipe/pipes/__init__.py

```diff
@@ -73,15 +73,27 @@ def update_or_create_package(project, package_data, codebase_resource=None):
     """
     Gets, updates or creates a DiscoveredPackage then returns it.
     Uses the `project` and `package_data` mapping to lookup and creates the
-    DiscoveredPackage using its Package URL as a unique key.
+    DiscoveredPackage using its Package URL and package_uid as a unique key.
     """
     purl_data = DiscoveredPackage.extract_purl_data(package_data)
+    package_uid = package_data.get("package_uid")
+    purl_data_and_package_uid = {
+        **purl_data,
+        "extra_data": {"package_uid": package_uid},
+    }
 
     try:
-        package = DiscoveredPackage.objects.get(project=project, **purl_data)
+        package = DiscoveredPackage.objects.get(
+            project=project,
+            **purl_data_and_package_uid
+        )
     except DiscoveredPackage.DoesNotExist:
         package = None
 
+    package_data = {
+        **package_data,
+        "extra_data": {"package_uid": package_uid},
```

Comment:
> The extra_data probably isn't the best place to keep the package_uids, so I'd like to know how we should handle package_uids in DiscoveredPackages. What is the best way to store package_uids on DiscoveredPackages?

@JonoYang Revisit this whole PR, you were right, using extra_data is not great for storing package_uid and it adds some complexity on the query side of things.
This value seems important enough to be its own field.
Now that the code is more stable, it's probably a good time to add a new field on the DiscoveredPackage model.
Would a simple `package_uid = models.UUIDField()` be good enough as a start?

Reply:
@tdruez 

> Would a simple package_uid = models.UUIDField() be good enough as a start?

the package_uid is a plain string not an actual bare UUID
It is made from a purl with a uuid as a qualifier, I would use something like 500 chars to account for other possible long qualifiers, or as long as a what we use for URL and purl fields

Reply:
Base on PackageURLMixin model, the purl value can be up to 1695 chars.
I would suggested 1024 for the package_uid to be a the safe side.

Reply:
perfect.